composer/datasets/hf_streaming.py

```diff
@@ -0,0 +1,243 @@
+# Copyright 2021 MosaicML. All Rights Reserved.
+
+import logging
+from dataclasses import dataclass
+from functools import partial
+from itertools import chain
+from typing import List, Optional
+
+import yahp as hp
+from torch.utils.data import IterableDataset
+
+from composer.core.types import Batch, DataSpec
+from composer.datasets.dataloader import DataloaderHparams
+from composer.datasets.hparams import DatasetHparams
+from composer.utils import dist
+
+log = logging.getLogger(__name__)
+
+
+def _split_dict_fn(batch: Batch, n_microbatches: int) -> List[Batch]:
+    if isinstance(batch, dict):
+        chunked = {k: v.chunk(n_microbatches) for k, v in batch.items()}
+        num_chunks = len(list(chunked.values())[0])
+        return [{k: v[idx] for k, v in chunked.items()} for idx in range(num_chunks)]
+    else:
+        raise ValueError(f'Expect batch from dataloader to be of type Dict[str, Tensor], but got {type(batch)}')
+
+
+CACHED_DATASET_SIZES = {"c4": {"en": {"train": (1024, 356317), "validation": (8, 45576)}}}
+
+
+@dataclass
+class HFStreamingDatasetHparams(DatasetHparams):
+    """Defines a generic dataset class for autoregressive and masked language models."""
+
+    name: str = hp.optional("Name of the dataset to load.", default=None)
+    config_name: Optional[str] = hp.optional(
+        "If required, the specific configuration of the dataset that you would like to use.", default=None)
+    split: str = hp.optional("What split of the dataset to use (e.g. 'train' or 'validation' or 'test')", default=None)
+    max_shards: int = hp.optional("Max number of shards, used to deterministically reduce dataset size.", default=-1)
```

Comment:
What would -1 mean?

Reply:
Removing `max_shards` as we are making this example C4 specific.